components/server/src/websocket-connection-manager.ts

```diff
@@ -85,4 +102,47 @@ export class WebsocketConnectionManager<C extends GitpodClient, S extends Gitpod
             this.servers.splice(index);
         }
     }
+}
+
+class GitpodJsonRpcConnectionHandler<T extends object> extends JsonRpcConnectionHandler<T> {
+    constructor(
+        readonly path: string,
+        readonly targetFactory: (proxy: JsonRpcProxy<T>, request?: object) => any,
+        readonly accessGuard: (request?: object) => FunctionAccessGuard
+    ) { 
+        super(path, targetFactory);
+    }
+
+    onConnection(connection: MessageConnection, request?: object): void {
+        const factory = new GitpodJsonRpcProxyFactory<T>(this.accessGuard(request));
+        const proxy = factory.createProxy();
+        factory.target = this.targetFactory(proxy, request);
+        factory.listen(connection);
+    }
+}
+
+class GitpodJsonRpcProxyFactory<T extends object> extends JsonRpcProxyFactory<T> {
+
+    constructor(protected readonly accessGuard: FunctionAccessGuard) { 
+        super();
+    }
+    
+    protected async onRequest(method: string, ...args: any[]): Promise<any> {
```

Comment:
fyi: there is also `onNotification` for methods started with `on` or `notify`. `GitpodServer` does not have such right now, maybe we can override this method and always throw just to be sure?